### DISCUSSION TITLE
Lidar Map 101

### DISCUSSION THREAD

**Discussion by krittin-kulrat** [UndefiedCse/Heat-Island#10] [General]
# Overview
This is what I got so far after researching on how to use Lidar Map. I'll update this post as more information I found on the internet.

# Terminology
## LAS file
**LAS** file is a type of file that store LiDAR points XYZ measurement, attributes and other metadata.

> I believe that LAS file is a raw data from LiDAR.

## LAZ file
**LAZ** file is a compressed LAS file
## Digital Elevation Models (DEM)
A numeric representation of the earth surface. Sometimes, DEM are stored in `.geotiff` or `.tiff` file. contain only coordinates and elevations.

> I think this is the data we need for our project. However, it might be harder to get this data.

### Digital Surface Model (DSM)
DEM that **includes** objects on the surface such as vegetation or buildings produced from the first return of laser signal.
### Digital Terrain Model (DTM)
DEM that **excludes** object on the surface meaning that it shows bear earth with vegetation and buildings removed. DEM is produced from the last return of laser signal.

# Database/Dataset
## [U.S. Geological Survey (USGS)](https://apps.nationalmap.gov/lidar-explorer/#)
This database is just an aggregation of LiDAR database in the U.S. Each database has its own resolution, metadata, and some area might has multiple LiDAR point maps.

## [Scottish Remote Sensing Portal](https://remotesensingdata.gov.scot/)
This database is a LiDAR dataset made by Scottish Government. This database provides LAZ, DSM, and DTM data.

## [GeoPackageData](https://www.geopackage.org/data.html)
Compilation of database for various country such as AUS, CAN, NZ, UK, and Non-governmental database.

> It looks like for Canada there DEM data available under "CDEM" and "CDSM" name. For UK, I only saw DTM data under "OS Terrain 50" name.

# Relevant Library
List of libraries that we might need for processing with LiDAR dataset.
## PyQGIS
QGIS is a user-friendly Open-Source Geographic Information System (GIS) licensed under the GNU General Public License. QGIS is an official project of the Open-Source Geospatial Foundation (OSGeo). It runs on Linux, Unix, Mac OSX, Windows and Android and supports numerous vector, raster, and database formats and functionalities.

## Rasterio
Rasterio reads and writes geospatial raster data. Rasterio 1.4 works with Python 3.9+, Numpy 1.21+, and GDAL 3.3+. Official binary packages for Linux, macOS, and Windows with most built-in format drivers plus HDF5, netCDF, and OpenJPEG2000 are available on PyPI.

[Rasterio Documentation](https://rasterio.readthedocs.io/)

> This library is the first library I saw when researching this topic. I think we have every high potential to use this library.

## Laspy
Laspy is a python library for reading, modifying and creating LAS LiDAR files.
Laspy is compatible with Python 3.8+.

[Laspy Documentation](https://laspy.readthedocs.io/en/latest/)

# Reference
[LiDAR Tutorial using QGIS](https://datasciencescotland.github.io/LiDAR-Tutorial-QGIS/#introduction)

[QGIS package on conda](https://anaconda.org/conda-forge/qgis)

[PyQGIS Documentation](https://docs.qgis.org/3.28/en/docs/pyqgis_developer_cookbook/index.html)
Good for implementing but mostly focus on their application QGIS

[QGIS Python API](https://qgis.org/pyqgis/master/index.html)
Actual documentation (Very hard to read)

[Rasterio](https://anaconda.org/conda-forge/rasterio)

[LASPY](https://anaconda.org/conda-forge/laspy/)

[GeoPackage](https://www.geopackage.org/)
An encoding standard. Grand father of everything above

**Comment by LilacHo** [comment]
Hi Krittin, thank you for compiling this information so thoroughly and coherently. Two comments: 
1. It seems like there is no existing package or github repository to convert Lidar data into csv format, right? I guess we can incorporate this as a piece of our project (like a self-defined function).
2. I used pyQGIS before but in windows not in Linux. However, I assume that @JornsenChao might have greater expertise in geodata. I do want to see a sample dataset ASAP; otherwise, I cannot keep myself from doubting if we can actually get height information or not. This is one task we need to do next step.

**Reply by krittin-kulrat** [reply]
Here is code from ChatGPT to convert `.geotif` and `.tif` files into `.csv`. As well as `.las` to `.csv`

<details><summary>Geotif to CSV</summary>
<p>

```python
import rasterio
import numpy as np
import pandas as pd

# Open the DEM file
dem_file = "your_dem_file.tif"  # Replace with the path to your DEM file
with rasterio.open(dem_file) as src:
    # Read the DEM data as a NumPy array
    dem_data = src.read(1)

    # Get the spatial information
    transform = src.transform
    crs = src.crs

    # Get the height (elevation) values and the corresponding coordinates
    heights = dem_data.flatten()
    x, y = np.meshgrid(np.arange(src.width), np.arange(src.height))
    x = x.flatten()
    y = y.flatten()

    # Transform pixel coordinates to geographic coordinates
    lon, lat = rasterio.transform.xy(transform, x, y)

    # Create a Pandas DataFrame to store the data
    data = {
        'Longitude': lon,
        'Latitude': lat,
        'Elevation': heights
    }

    df = pd.DataFrame(data)

# Save the data as a CSV file
csv_file = "dem_data.csv"  # Choose a name for your CSV file
df.to_csv(csv_file, index=False)

print(f"DEM data has been saved to {csv_file}")

``` 


</p>
</details> 

<details><summary>LAS to CSV</summary>
<p>

```python
import laspy
import csv

# Specify the input LAS file
input_las_file = 'input_file.las'

# Specify the output CSV file
output_csv_file = 'output_file.csv'

# Open the LAS file
with laspy.file.File(input_las_file, mode='r') as las_file:
    # Extract the point records from the LAS file
    points = las_file.points

    # Extract the desired data: X, Y, Z coordinates and classification
    X = las_file.x
    Y = las_file.y
    Z = las_file.z
    classifications = las_file.classification

# Write the data to a CSV file
with open(output_csv_file, 'w', newline='') as csvfile:
    fieldnames = ['X', 'Y', 'Z', 'Classification']
    writer = csv.DictWriter(csvfile, fieldnames=fieldnames)

    writer.writeheader()
    for i in range(len(points)):
        writer.writerow({'X': X[i], 'Y': Y[i], 'Z': Z[i], 'Classification': classifications[i]})

print(f'Done writing data to {output_csv_file}')
```

</p>
</details> 
From the look of it, I don't think it's going to consume our time that much but might need to tinker it a little bit

For height information, we can assume that elevation from DSM data is rooftop and subtract them by elevation from DTM data to get building height. We can also average elevation in the area around it (~10-20m may be) so it's not sensitive to noise in the data.

**Comment by LilacHo** [comment]
Seattle city boundary: https://data-seattlecitygis.opendata.arcgis.com/datasets/c5f3575dd7d545ada27064c74ac74f52_0/explore?location=47.622905%2C-122.278830%2C11.48
